ccl/changefeedccl: TestChangefeedLaggingSpanCheckpointing failed

ccl/changefeedccl.TestChangefeedLaggingSpanCheckpointing [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/19737203?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/19737203?buildTab=artifacts#/) on release-25.2 @ [3247f4deec8bf8aa9cc8c4f7b998192abc1bc935](https://github.com/cockroachdb/cockroach/commits/3247f4deec8bf8aa9cc8c4f7b998192abc1bc935):


```
=== RUN   TestChangefeedLaggingSpanCheckpointing
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/51fe861426dc77645f2622be784d4d3c/logTestChangefeedLaggingSpanCheckpointing4145386488
    test_log_scope.go:76: use -show-logs to present logs inline
    changefeed_test.go:2400: cursor: 1747767907.290948786,0
    changefeed_test.go:2444: found checkpoint: map[1747767907.290948786,0:/Table/106/1{-/51}]
    changefeed_test.go:2444: found checkpoint: map[1747767907.290948786,0:/Table/106/1{-/51}]
    changefeed_test.go:2444: found checkpoint: map[1747767907.290948786,0:/Table/106/1{-/51}]
    changefeed_test.go:2444: found checkpoint: map[1747767908.058023825,0:/Table/106/1/{351-451}, /Table/106/1/{551-651}, /Table/106/1/{701-801}, /Table/106/1/{851-901} 1747767908.258315359,0:/Table/106/1/9{01-51} 1747767911.083419877,0:/Table/106/1/{51-101}, /Table/106/1/{151-201}]
    changefeed_test.go:2460: found checkpoint: map[1747767908.058023825,0:/Table/106/1/{351-451}, /Table/106/1/{551-651}, /Table/106/1/{701-801}, /Table/106/1/{851-901} 1747767908.258315359,0:/Table/106/1/9{01-51} 1747767911.083419877,0:/Table/106/1/{51-101}, /Table/106/1/{151-201}]
    changefeed_test.go:2518: 
        	Error Trace:	pkg/ccl/changefeedccl/changefeed_test.go:2518
        	Error:      	Received unexpected error:
        	            	rangefeed for span /Table/106/{1/951-2} expected to start @1747767907.290948786,0, started @1747767911.685473379,0 instead
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestChangefeedLaggingSpanCheckpointing.func4.1
        	            	  | 	pkg/ccl/changefeedccl/changefeed_test.go:2478
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestChangefeedLaggingSpanCheckpointing.func4
        	            	  | 	pkg/ccl/changefeedccl/changefeed_test.go:2493
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed.rangefeedFactory.Run.func1
        	            	  | 	pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go:111
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed.rangefeedFactory.Run.Group.GoCtx.func7
        	            	  | 	pkg/util/ctxgroup/ctxgroup.go:199
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	external/org_golang_x_sync/errgroup/errgroup.go:78
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_arm64.s:1223
        	            	Wraps: (2) rangefeed for span /Table/106/{1/951-2} expected to start @1747767907.290948786,0, started @1747767911.685473379,0 instead
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestChangefeedLaggingSpanCheckpointing
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/51fe861426dc77645f2622be784d4d3c/logTestChangefeedLaggingSpanCheckpointing4145386488
--- FAIL: TestChangefeedLaggingSpanCheckpointing (7.05s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedLaggingSpanCheckpointing.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-50864

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -2367,7 +2367,6 @@ func TestChangefeedSchemaChangeNoBackfill(t *testing.T) {
 func TestChangefeedLaggingSpanCheckpointing(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-	rnd, _ := randutil.NewPseudoRand()
 
 	s, db, stopServer := startTestFullServer(t, feedTestOptions{})
 	defer stopServer()
@@ -2383,6 +2382,9 @@ func TestChangefeedLaggingSpanCheckpointing(t *testing.T) {
   INSERT INTO foo (key) SELECT * FROM generate_series(1, 1000);
   ALTER TABLE foo SPLIT AT (SELECT * FROM generate_series(1, 1000, 50));
   `)
+
+	fooDesc := desctestutils.TestingGetPublicTableDescriptor(s.DB(), s.Codec(), "d", "foo")
+	tableSpan := fooDesc.PrimaryIndexSpan(s.Codec())
 
 	// Checkpoint progress frequently, allow a large enough checkpoint, and
 	// reduce the lag threshold to allow lag checkpointing to trigger
@@ -2403,25 +2405,29 @@ func TestChangefeedLaggingSpanCheckpointing(t *testing.T) {
 	// Rangefeed will skip some of the checkpoints to simulate lagging spans.
 	var laggingSpans roachpb.SpanGroup
 	nonLaggingSpans := make(map[string]int)
-	var numLagging, numNonLagging int
-	knobs.FeedKnobs.ShouldSkipCheckpoint = func(checkpoint *kvpb.RangeFeedCheckpoint) bool {
+	var numSpans int
+	knobs.FeedKnobs.ShouldSkipCheckpoint = func(checkpoint *kvpb.RangeFeedCheckpoint) (skip bool) {
+		// Skip spans for the whole table.
+		if checkpoint.Span.Equal(tableSpan) {
+			return true
+		}
 		// Skip spans that we already picked to be lagging.
 		if laggingSpans.Encloses(checkpoint.Span) {
-			return true /* skip */
+			return true
 		}
-		// Skip additional updates for some non-lagging spans so that we can
-		// have more than one timestamp in the checkpoint.
+		// Skip additional updates for every 3rd non-lagging span so that we have
+		// a few spans lagging at a second timestamp above the cursor.
 		if i, ok := nonLaggingSpans[checkpoint.Span.String()]; ok {
 			return i%3 == 0
 		}
-		// Ensure we have a few spans that are lagging at the cursor.
-		if numLagging == 0 || (numLagging < 5 && rnd.Int()%3 == 0) {
+		numSpans++
+		// Skip updates for every 3rd span so that we have a few spans lagging
+		// at the cursor.
+		if numSpans%3 == 0 {
 			laggingSpans.Add(checkpoint.Span)
-			numLagging++
-			return true /* skip */
+			return true
 		}
-		nonLaggingSpans[checkpoint.Span.String()] = numNonLagging
-		numNonLagging++
+		nonLaggingSpans[checkpoint.Span.String()] = len(nonLaggingSpans) + 1
 		return false
 	}
 
@@ -2470,10 +2476,16 @@ func TestChangefeedLaggingSpanCheckpointing(t *testing.T) {
 		}
 	}
 
-	var rangefeedStarted bool
+	var rangefeedStartedOnce bool
 	var incorrectCheckpointErr error
 	knobs.FeedKnobs.OnRangeFeedStart = func(spans []kvcoord.SpanTimePair) {
-		rangefeedStarted = true
+		// We only need to check the first rangefeed restart since
+		// any additional restarts (likely due to transient errors)
+		// may be using newer span-level checkpoints than the one
+		// we saved after the last pause.
+		if rangefeedStartedOnce {
+			return
+		}
 
 		setErr := func(stp kvcoord.SpanTimePair, expectedTS hlc.Timestamp) {
 			incorrectCheckpointErr = errors.Newf(
@@ -2495,6 +2507,8 @@ func TestChangefeedLaggingSpanCheckpointing(t *testing.T) {
 				}
 			}
 		}
+
+		rangefeedStartedOnce = true
 	}
 	knobs.FeedKnobs.ShouldSkipCheckpoint = nil
 
@@ -2514,7 +2528,7 @@ func TestChangefeedLaggingSpanCheckpointing(t *testing.T) {
 	waitForJobState(sqlDB, t, jobID, jobs.StatePaused)
 	// Verify the rangefeed started. This guards against the testing knob
 	// not being called, which was happening in earlier versions of the code.
-	require.True(t, rangefeedStarted)
+	require.True(t, rangefeedStartedOnce)
 	// Verify we didn't see incorrect timestamps when resuming.
 	require.NoError(t, incorrectCheckpointErr)
 }

